core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -578,7 +580,9 @@ void performOnPrimary(final ShardRouting primary, final ShardIterator shardsIt)
                 PrimaryOperationRequest por = new PrimaryOperationRequest(primary.id(), internalRequest.concreteIndex(), internalRequest.request());
                 Tuple<Response, ReplicaRequest> primaryResponse = shardOperationOnPrimary(observer.observedState(), por);
                 logger.trace("operation completed on primary [{}]", primary);
-                replicationPhase = new ReplicationPhase(shardsIt, primaryResponse.v2(), primaryResponse.v1(), observer, primary, internalRequest, listener, indexShardReference);
+                // TODO: set a default timeout
+                TimeValue shardFailedTimeout = settings.getAsTime(REPLICATION_ACTION_SHARD_FAILURE_TIMEOUT, null);
```

Comment:
in generation, this should be a variable that's resolved and cached. We shouldn't do settings lookups with each op.